### PR TITLE
Feature/#21/i_jepa_train

### PR DESCRIPTION
**Pull request by zassou65535**
## 概要

I-JEPAの学習用スクリプトとそのテスト、設定ファイルの追加

## Submit前の確認項目

- [x] タイトルは一目でわかるようにし、説明文は PR を簡潔に説明するようにしましたか?
- [x] あなたの PR が、異なる変更を束ねたものではなく、ひとつのことだけを行うものであることを確認しましたか?
- [x] この PR で導入されたすべての変更点をリストアップしましたか?
- [x] PR を`make run`コマンドでローカルでテストしましたか?

### CODE REVIEW

**Comment by Geson-anko** [comment]
@zassou65535 Trainerの `get_tranining_model`は `ModelNames`を指定するようにしてほしい。諸事情あって、typingが`str`になってるけど、将来的に修正する予定

https://github.com/MLShukai/ami-vconf24/pull/47/commits/cfbb05aee79733473c569cd90480febeb7538de0

File: ami/trainers/i_jepa_trainer.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+                masks_for_predictor = [masks.to(self.device) for masks in masks_for_predictor]
+                optimizer.zero_grad()
+
+                # target encoder
+                with torch.no_grad():
+                    latent_from_target_encoder = target_encoder(image_batch)
+                    latent_from_target_encoder = torch.nn.functional.layer_norm(
+                        latent_from_target_encoder,
+                        (latent_from_target_encoder.size(-1),),
+                    )  # normalize over feature-dim
```
Comment (Geson-anko) [inline ami/trainers/i_jepa_trainer.py:118]: `IJEPAEncoder`の`forward`の最後でLayerNormをかけていたから意味がないんじゃないかな。 強いていうなならbiasがないことくらい？ ContextEncoderとTargetEncoderの処理フローは一致していないと困るんじゃないかな。
```suggestion
```

File: ami/trainers/i_jepa_trainer.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        self.predictor: ModelWrapper[IJEPAPredictor] = self.get_training_model(ModelNames.I_JEPA_PREDICTOR)
+        self.target_encoder: ModelWrapper[IJEPAEncoder] = self.get_training_model(ModelNames.I_JEPA_TARGET_ENCODER)
+
+        # Since the model is swapped between the inference and training threads each time it is trained,
+        # the model and optimizer are built within the `train()` method.
+        # The following is the initial state generation of the optimizer.
+        self.optimizer_state = self.partial_optimizer(
+            itertools.chain(self.context_encoder.parameters(), self.predictor.parameters())
+        ).state_dict()
+
```
Comment (Geson-anko) [inline ami/trainers/i_jepa_trainer.py:75]: これは実装を確認して欲しいんだけど、TargetEncoderとContextEncoderって同じパラメータからスタートすると思う。ただ、ちょっと AMI Systemの実装的に、instantiateをそれぞれに対して呼び出すことになりそうで、初期パラメータの同一性を保証できない。だからon_model_wrappers_dict_attachedの中でパラメータのコピーをして欲しい。

また、コピーする時には Target EncoderからCortextEncoderにコピーすることを注意してほしい。
 Target Encoderは 推論スレッドと学習スレッドに**2重に** 存在していて、TargetEncoderの方のパラメータを変えちゃうと同期処理を手動で呼ばなければいけなくなってしまうんだ。
```suggestion
        for context_encoder_param, target_encoder_param in zip(self.context_encoder.paramters(), self.target_encoder.parameters()):
            context_encoder_param.copy_(target_encoder_param)
```

File: tests/trainers/test_i_jepa_trainer.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+    @pytest.fixture
+    def model_wrappers_dict(
+        self, device: torch.device, i_jepa_encoder: IJEPAEncoder, i_jepa_predictor: IJEPAPredictor
+    ) -> ModelWrappersDict:
+        d = ModelWrappersDict(
+            {
+                "i_jepa_context_encoder": ModelWrapper(i_jepa_encoder, device, False),
+                "i_jepa_predictor": ModelWrapper(i_jepa_predictor, device, False),
+                "i_jepa_target_encoder": ModelWrapper(copy.deepcopy(i_jepa_encoder), device, False),
```
Comment (Geson-anko) [inline tests/trainers/test_i_jepa_trainer.py:109]: ModelNamesを使用しよう。また、target encoderは agentの推論スレッドでも用いられるので`has_inference`を`True`にしよう。
```suggestion
                ModelNames.I_JEPA_CONTEXT_ENCODER: ModelWrapper(i_jepa_encoder, device, False),
                ModelNames.I_JEPA_PREDICTOR: ModelWrapper(i_jepa_predictor, device, False),
                ModelNames.I_JEPA_TARGET_ENCODER: ModelWrapper(copy.deepcopy(i_jepa_encoder), device, True),
```

File: tests/trainers/test_i_jepa_trainer.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            }
+        )
+        assert not (
+            id(d["i_jepa_context_encoder"].model) == id(d["i_jepa_target_encoder"].model)
+        ), "context_encoder and target_encoder must be allocated in memory as separate entities."
```
Comment (Geson-anko) [inline tests/trainers/test_i_jepa_trainer.py:114]: テストのコンテキスト的に要らない。
```suggestion
```

**Review by Geson-anko — CHANGES_REQUESTED** [review]
@zassou65535 実装ありがとう。AMI Systemの内部構造に関連した修正が多いから、よくわからなかったら気軽に質問してほしい。

File: ami/trainers/i_jepa_trainer.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        return self.image_data_user.buffer.new_data_count >= self.minimum_new_data_count
+
+    def train(self) -> None:
+        # define models
+        context_encoder = self.context_encoder.model
+        predictor = self.predictor.model
+        target_encoder = self.target_encoder.model
+        assert not (
+            id(context_encoder) == id(target_encoder)
+        ), "context_encoder and target_encoder must be allocated in memory as separate entities."
```
[the remark this answers is not in the document]
Comment (zassou65535) [reply ami/trainers/i_jepa_trainer.py]: https://github.com/MLShukai/ami-vconf24/pull/47/commits/4bac76bcc0434d77506e0c83c82f509b44bdc589 で修正

File: ami/trainers/i_jepa_trainer.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        self.minimum_new_data_count = minimum_new_data_count
+
+    def on_data_users_dict_attached(self) -> None:
+        self.image_data_user: ThreadSafeDataUser[RandomDataBuffer] = self.get_data_user(BufferNames.IMAGE)
+
+    def on_model_wrappers_dict_attached(self) -> None:
+        self.context_encoder: ModelWrapper[IJEPAEncoder] = self.get_training_model(ModelNames.I_JEPA_CONTEXT_ENCODER)
+        self.predictor: ModelWrapper[IJEPAPredictor] = self.get_training_model(ModelNames.I_JEPA_PREDICTOR)
+        self.target_encoder: ModelWrapper[IJEPAEncoder] = self.get_training_model(ModelNames.I_JEPA_TARGET_ENCODER)
+
```
[the remark this answers is not in the document]
Comment (zassou65535) [reply ami/trainers/i_jepa_trainer.py:67]: https://github.com/MLShukai/ami-vconf24/pull/47/commits/4bac76bcc0434d77506e0c83c82f509b44bdc589 で修正

File: tests/trainers/test_i_jepa_trainer.py (hunk 4 above)
Comment (zassou65535) [reply tests/trainers/test_i_jepa_trainer.py]: https://github.com/MLShukai/ami-vconf24/pull/47/commits/4bac76bcc0434d77506e0c83c82f509b44bdc589 で修正

File: tests/trainers/test_i_jepa_trainer.py (hunk 3 above)
Comment (zassou65535) [reply tests/trainers/test_i_jepa_trainer.py]: https://github.com/MLShukai/ami-vconf24/pull/47/commits/313328dcabae403ee6e5357837e2eaae3a439d7f で変更

File: ami/trainers/i_jepa_trainer.py (hunk 2 above)
Comment (zassou65535) [reply ami/trainers/i_jepa_trainer.py:74]: https://github.com/MLShukai/ami-vconf24/pull/47/commits/9845eabb6a827596aa155f3c822283c719211230 で追加

File: ami/trainers/i_jepa_trainer.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        optimizer.load_state_dict(self.optimizer_state)
+        # prepare about dataset
+        dataset = self.image_data_user.get_dataset()
+        dataloader = self.partial_dataloader(dataset=dataset, collate_fn=self.i_jepa_mask_collator)
+
+        for _ in range(self.max_epochs):
+            for batch in dataloader:
+                (image_batch, masks_for_context_encoder, masks_for_predictor) = batch
+                assert len(image_batch) == 1  # must be fixed in another PR
+                image_batch = image_batch[0]  # must be fixed in another PR
```
[the remark this answers is not in the document]
Comment (zassou65535) [reply ami/trainers/i_jepa_trainer.py:107]: これ調べてみたけど、`IJEPAMultiBlockMaskCollator`の`__call__`メソッドの引数`images`に、本来`List[torch.Tensor]`型想定の変数が渡されるはずだったところに、学習実行時には`List[List[torch.Tensor]]`が渡されてしまってたのが原因っぽかった
イメージとしては、例えばバッチサイズ2なら本来`[torch.Tensor, torch.Tensor]`みたいなListが入ってくるはずなんだけど、`[[torch.Tensor], [torch.Tensor]]`が入ってきてしまってる感じ(謎に`.unsqueeze(dim=1)`のような何かがListにかかってる感じ)
修正するなら変更が`IJEPAMultiBlockMaskCollator`にもまたがってくるかもだから、これは別プルリクで修正にしたいかも

[the remark this answers is not in the document]
Comment (Geson-anko) [reply ami/trainers/i_jepa_trainer.py]: @zassou65535 これはRandomDataBufferの内部でtorch.utils.data.TensorDatasetクラスを用いているからなんだよね。
修正しておいた
https://github.com/MLShukai/ami-vconf24/pull/47/commits/44b3fd57910ee5140c79da1ef5346754f0395cec

**Review by Geson-anko — APPROVED** [review]
@zassou65535 実装ありがとう。collatorの方は修正しておいた。
機能の大幅な修正でない限り、そのファイルを編集する影響者がでないから、細かな修正はそのPRでやってOK. だってそのPRが上手く動かないから修正するから、一緒になっていた方が修正の意図を拾いやすい。

**Review by Geson-anko — APPROVED** [review]
Approveします
